proof_of_concept/rest/ddm_site.py

```diff
@@ -34,6 +34,47 @@
 logger = logging.getLogger(__name__)
 
 
+def _request_url(request: Request) -> str:
+    """Obtain the URL for the current request.
+
+    This URL-quotes the path part of the URL, encoding everything
+    except /.
+
+    If we are running behind a reverse proxy, then what gunicorn thinks
+    is our external location is actually an internal one. This function
+    returns the actual external URL of the current request, excluding
+    the parameters. If we're not behind a reverse proxy, then that's
+    just the normal scheme + host + path, if we are behind a reverse
+    proxy then the following headers must be set:
+
+        X-Forwarded-Proto: The protocol used, http or https
+        X-Forwarded-Host: The outside host part
+        X-Forwarded-Path: The outside path
+
+    Note that X-Forwarded-Path is very non-standard. Because of our
+    split API setup with two different reverse proxies, the external
+    and internal paths aren't the same, so we need the external path
+    as well to be able to generate correct client-resolvable URLs.
```

Comment:
This seems complicated and non-standard, but you documented it very well. I am curious if there is a simpler solution to this, or better: move things around a bit so we don't have to do this. I have absolutely 0 ideas for alternatives, but I just wanted to see if you do. Otherwise this is nicely documented and a good solution!

Reply:
One option would be to run separate WSGI servers for the internal and external APIs. They would have to communicate with the core however, probably through REST, and then you have the same set-up as we have now except you've implemented your HTTP reverse proxy from scratch. So that doesn't help.

A real alternative would be a data-centric architecture with e.g. a PostgreSQL database containing policies, assets, jobs and replicas, and two WSGI servers (one for each API) talking to it, and to each other via it. That would work and be a more standard design, and it would add the transactional persistent store that would be needed anyway in a production setting. It's a fair amount of work to implement though. Maybe in the future...

Reply:
Fair enough! It is clear that you considered alternatives!

Reply:
Yes. There's also the standard `Forwarded` header as a potential alternative to `X-Forwarded-Proto` and `X-Forwarded-Host`, but it's quite a bit more complex and it doesn't include the path either, so it doesn't solve the whole problem and is more difficult to parse. So I went with this.